sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -185,14 +185,15 @@ uint8_t *coap_find_option(const coap_pkt_t *pkt, unsigned opt_num)
  * opt_len[out]   length of option value
  *
  * return         next byte after option header, usually the option value
- * return         NULL if initial pkt_pos is past options
+ * return         NULL if initial pkt_pos is payload marker or past options
  */
 static uint8_t *_parse_option(const coap_pkt_t *pkt,
                               uint8_t *pkt_pos, uint16_t *delta, int *opt_len)
 {
     uint8_t *hdr_end = pkt->payload;
 
-    if (pkt_pos == hdr_end) {
+    if ((pkt_pos >= hdr_end)
+            || (((pkt_pos + 1) == hdr_end) && (*pkt_pos == 0xFF))) {
```

Comment:
I wonder if this can be simplified to `(pkt_pos >= hdr_end) || (*pkt_pos == 0xFF)` because `hdr_end = pkt->payload` and `0xFF` is the payload marker so if `pkt_pos == 0xFF` then `pkt_pos + 1` should be `hdr_end`.

Reply:
and hence we don't need the extra check verifying `pkt_pos + 1 == hdr_end`

Reply:
can be optimised later, if needed at all

Reply:
Thanks for pushing this PR forward, @smlng. I will look and comment later this week before your point gets lost in my memory.

Reply:
> so if pkt_pos == 0xFF then pkt_pos + 1 should be hdr_end

I agree that it is worthwhile to remove the extra check on position. It provides a tighter guarantee on the structure of options. I'll create a cleanup PR. Thanks!